crates/sui-network/src/discovery/mod.rs

```diff
@@ -70,6 +77,36 @@ pub struct NodeInfo {
     pub access_type: AccessType,
 }
 
+impl NodeInfo {
+    fn sign(self, keypair: &NetworkKeyPair) -> SignedNodeInfo {
+        let msg = bcs::to_bytes(&self).expect("BCS serialization should not fail");
+        let sig = keypair.sign(&msg);
+        SignedNodeInfo::new_from_data_and_sig(self, sig)
+    }
+}
+
+pub type SignedNodeInfo = Envelope<NodeInfo, Ed25519Signature>;
+
+pub type VerifiedSignedNodeInfo = VerifiedEnvelope<NodeInfo, Ed25519Signature>;
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+pub struct NodeInfoDigest(Digest);
+
+impl NodeInfoDigest {
+    pub const fn new(digest: [u8; 32]) -> Self {
+        Self(Digest::new(digest))
+    }
+}
+
+impl Message for NodeInfo {
+    type DigestType = Ed25519Signature;
+    const SCOPE: IntentScope = IntentScope::DiscoveryPeers;
+
+    fn digest(&self) -> Self::DigestType {
+        unreachable!("NodeInfoDigest is not used today")
+    }
+}
```

Comment:
Why are we adding this if its unreachable? Can we instead just have a SignedNodeInfo that is:

```
struct SignedNodeInfo {
    info: NodeInfo,
    signature: Ed25519Signature,
}
```
And just do a sig check before we insert it. into the hashmap and not need to use the envelope/verified types since we don't seem to be leveraging the Message impl with it anyway. (I don't necessarily think we even need a separate signed/verified type either since this is much less risky than txns/checkpoints for example

Reply:
I think the envelope/verified abstractions are a good fit here. We use this pattern elsewhere for similar reasons, engaging the type system to help make sure that we don't accidentally store unverified NodeInfo. It's true I didn't implement `digest`, mostly since we wouldn't have any use for it, other than to sign on that instead of the BCS serialization of the message. I copied the "unimplemented `digest`" pattern from the existing `BridgeAction` in the codebase, but could def implement `digest` and use it for the signing if we think that's cleaner. I am inclined to still use Envelope/VerifiedEnvelope rather than a struct containing a signature field for consistency with other parts of the codebase. But I def can see your argument that it might be slightly overkill.

Reply:
We sign the bcs serialization of the message and not its digest. I figured the extra hashing didn't really add anything but could change to implement digest on the bcs-serialized message and then sign the digest instead, if you think that's better?